### PR TITLE
[ops] feat: add Ascend Triton clamped SwiGLU

### PR DESCRIPTION
**Pull request by WenzheWang**
### What does this PR do?

Adds a forward/backward Ascend Triton implementation of DeepSeek-V4's clamped SwiGLU activation and wires it into VeOmni's `fused_npu` grouped-GEMM MoE path.

- Use the Ascend Triton kernel only when the installed Triton runtime exposes the actual `libtriton.ascend` backend.
- Preserve the original eager `chunk` / `clamp` / `F.silu` / multiply forward-and-backward path when Triton is absent or when only vanilla Triton is installed.
- Preserve `torch_npu.npu_swiglu` for models without `swiglu_limit`.
- Keep the Triton kernel import lazy so bare and legacy NPU environments retain their existing training path without gaining a required dependency.
- Keep DeepSeek-V4 NPU tests on the production `fused_npu` route, and add capability, fallback, dispatch, numerical, boundary-gradient, tail-shape, empty-input, and validation coverage.
- Document the numerical difference between the Triton FP32-compute path and the input-dtype eager fallback.

No exact duplicate was found in searches for [`swiglu_limit`](https://github.com/ByteDance-Seed/VeOmni/pulls?q=is%3Apr+is%3Aopen+swiglu_limit) or [`npu_group_gemm`](https://github.com/ByteDance-Seed/VeOmni/pulls?q=is%3Apr+is%3Aopen+npu_group_gemm). #1025 changes grouped-GEMM launch bounds and #896 changes the Ascend EP dispatcher; neither implements this clamped activation kernel.

The kernel was authored from VeOmni's existing eager clamped-SwiGLU formula and local Triton activation utilities; no third-party kernel code was copied.

### Checklist Before Starting

- [x] Searched for related PRs/issues and documented adjacent work above.
- [x] PR title follows `[{modules}] {type}: {description}`.

### Test

Current head: `7a65956b82faaecea8ec634b29e58a7de0c27eb7`.

Current-head CPU-collectible validation:

```bash
make quality

python -m pytest -q -p no:cacheprovider \
  tests/models/test_deepseek_v4_liger_ops.py

python -m pytest -q -p no:cacheprovider \
  tests/models/test_deepseek_v4_fused_moe.py::test_deepseek_v4_test_overrides_follow_active_device

python -m pytest -q -p no:cacheprovider --collect-only \
  tests/ops/test_npu_kernels.py

python tests/special_sanity/check_device_api_usage.py --directory tests
git diff --check
```

Results:

- Repository-locked Ruff quality: passed (`590 files already formatted`).
- DeepSeek-V4 dispatch, eager fallback forward/backward, vanilla-Triton capability rejection, and Ascend-Triton dispatch contracts: `8 passed, 3 skipped`.
- Active-device production override contract: `2 passed`.
- NPU suite collection: `31 tests collected`, including all 7 clamped-SwiGLU hardware cases.
- Device-API scan and `git diff --check`: passed.
- Independent two-pass review: `safe`; no actionable code, test, or documentation findings remain.

Ascend 910B2 validation was previously completed at old head `4fed50cf9ce1361a7faa6f44699ae6b1c2513eb9`. It remains useful evidence for the unchanged Triton kernel implementation, but is intentionally not presented as current-head evidence:

- Hardware-independent dispatch/dependency contracts: `2 passed`.
- Clamped SwiGLU NPU numerical kernel: `7 passed, 24 deselected`, covering fp16/bf16 forward and backward parity, clamp-boundary gradients, empty input, and invalid-shape handling.
- DeepSeek-V4 `fused_npu` forward/backward integration: `1 passed`.
- Environment: Ascend 910B2, Python 3.12.13, PyTorch 2.10.0, torch_npu 2.10.0.post5.dev20260831, Triton 3.2.0.
- Raw evidence: `/data/wenzhe/veomni-pr1104-npu-20260902T1821` on host 225.

The earlier full upstream GPU/NPU CI pass was for head `ab5ab3cc`. CI for the current head must complete again before merge and is tracked separately from the old-head manual hardware evidence.

### API and Usage Example

No public API or configuration key changes. Existing DeepSeek-V4 configurations that select `fused_npu` and provide `swiglu_limit` use Ascend Triton when the backend is present, and otherwise preserve the original eager activation. Other NPU MoE models continue to use `torch_npu.npu_swiglu`.

VeOmni's product-based Ascend images install and verify `triton-ascend`. Bare and legacy environments do not need it for correctness; installing a release compatible with the CANN and `torch_npu` stack enables the fused activation.

### Design & Code Changes

- The kernel flattens leading dimensions into rows and tiles the post-split hidden dimension with masked 1024-element blocks, so non-multiple tail sizes are supported.
- Forward clamps the gate to `max=limit`, clamps the up projection to `[-limit, limit]`, and computes `SiLU(gate) * up` in FP32 before storing the input dtype.
- Backward applies the same clamp semantics, including PyTorch-compatible inclusive gradients at the exact boundaries, and writes both gate/up gradients.
- `_swiglu` remains the production dispatch owner: `None` selects `torch_npu.npu_swiglu`; a numeric limit selects Ascend Triton only when `libtriton.ascend` exists and otherwise selects the eager fallback.
- Both non-EP and EP grouped-GEMM paths use that same activation dispatcher.
- FP16 kernel-vs-eager comparison uses a tighter `5e-3` tolerance. BF16 retains `2e-2` because the Triton kernel computes in FP32 while eager preserves input dtype; exact zero-gradient assertions independently guard both clamp masks, and the FP32 boundary test remains tight.

### Checklist Before Submitting

- [x] Read the Contribute Guide.
- [x] Applied `make quality`.
- [x] Added/updated documentation.
- [x] Added CPU-collectible capability, fallback, and production-dispatch contracts.
- [x] Kept the hardware NPU kernel-vs-eager numerical and boundary-gradient tests.
- [ ] Current-head upstream GPU/NPU CI passes.
- [ ] Current-head Ascend Triton numerical evidence is available from upstream CI or a 910B2 rerun.

### CODE REVIEW

**Comment by WenzheWang** [comment]
@Luosuu @Coach257 This PR is synchronized with current `main` `d697023c` at head
`4fed50cf`. The latest CodeRabbit fixture-lifecycle finding is fixed, and all three review threads
are resolved. Repository-locked quality and the CPU contracts pass. Exact-head Ascend 910B2
validation on host 225 passed 2 CPU dispatch/dependency contracts, 7 NPU numerical kernel cases,
and 1 DeepSeek-V4 `fused_npu` forward/backward integration case, with all NPUs clear afterward.
The PR body contains the complete source/runtime identities and results. Could you please review
the current head and approve the maintainer-gated workflows when convenient? Thanks.

**Review by Luosuu — DISMISSED** [review]

**Comment by Luosuu** [comment]
@WenzheWang triggered CI for you. feel free to at me for merge or trigger CI another time

**Review by Luosuu — COMMENTED** [review]
please fix CI

**Comment by WenzheWang** [comment]
Thanks — fixed in `ab5ab3cc`.

The root cause was that the generic DeepSeek-V4 NPU unit/E2E tests selected `fused_npu` solely from the active device, although the default NPU CI image does not include the optional `triton` package. The tests now select `fused_npu` only when Triton is available and otherwise keep the eager baseline. Dedicated clamped-SwiGLU kernel tests skip only when Triton is absent, so real kernel failures remain visible. I also replaced the raw CUDA literal flagged by `device-api-check` with the repository device abstraction.

Local validation passed: Ruff check/format, the full device-API scan, focused regression tests (`4 passed`), and independent review. The new current-head CI is now queued/running.

**Comment by WenzheWang** [comment]
@Luosuu The CI fix is complete on the current head `ab5ab3cc9eb7c6e150b1762f5a74a118052b2503`. All nine current-head workflows now pass, including GPU/NPU unit and E2E tests, format, title, doc-path, and device-API checks. CodeRabbit, Read the Docs, and CLA are also green, and all three review threads are resolved/outdated.

Could you please re-review and merge when convenient? Thanks.

**Review by Coach257 — COMMENTED** [review]
Please keep the original eager clamp as the fallback when `triton-ascend` is missing, instead of raising. That path was already a full **fwd+bwd** training path (`chunk` / `clamp` / `F.silu` / mul, sitting between two `GmmFunction` grouped-GEMMs). DeepSeek-V4 always passes `swiglu_limit`, and NPU default MoE is `fused_npu`, so a V4 job on a bare NPU image could already train. Tests currently hide this by selecting `eager` when `triton` is absent; production configs do not.

On the new kernel itself: this is an Ascend Triton activation inside `fused_npu`, not GPU `fused_triton`. The formula matches eager (gate `max=limit`, up `[-limit, limit]`, inclusive clamp grads), but the numerics are not the same dtype path:

- Triton computes SiLU in FP32, then stores to fp16/bf16
- the old eager helper runs `clamp` + `F.silu` in the tensor dtype
- V4’s official shared-expert eager path upcasts with `.float().clamp()` before SiLU

So they can align, but not bit-exact in mixed precision.

CI coverage for that alignment is also thin:

- `TestNPUClampedSwiGLU.test_forward_backward_matches_eager` is the only direct kernel-vs-eager check, and it is skipped unless both NPU **and** `triton` are present. Default GitHub CPU CI never runs it.
- fp16/bf16 uses `rtol=2e-2, atol=2e-2`, which is too loose to catch a wrong clamp or a missing boundary mask. The tight `1e-5` case is fp32-only.
- CPU tests only assert dispatch / missing-package errors; they do not check values.
- `test_models_patch[deepseek_v4]` can exercise `fused_npu` vs HF eager on NPU, but that is a full-model check, and this fork’s NPU workflow was still waiting on maintainer approval.

Could you (1) restore eager as fallback, and (2) keep a hardware NPU test that compares the Triton kernel against that same eager formula with a tighter tolerance (or at least document the expected FP32-accum vs bf16 eager gap)? Without that, it is hard to tell from CI that the new kernel is mathematically aligned, not just wired in.

---

**Comment by WenzheWang** [comment]
@Coach257 Thanks — addressed in `7a65956b`.

- Restored the original eager clamp / `F.silu` / multiply forward-and-backward path whenever Ascend Triton is unavailable.
- Capability detection now checks the actual `libtriton.ascend` backend, so vanilla Triton also falls back correctly.
- DeepSeek-V4 NPU tests stay on the production `fused_npu` route instead of hiding the fallback by selecting eager.
- The direct NPU kernel comparison now uses `5e-3` for FP16. BF16 retains `2e-2` with the FP32-compute versus input-dtype eager difference documented, while exact zero-gradient assertions independently guard both clamp masks and the FP32 boundary test remains tight.
- Added CPU regressions for eager forward/backward fallback, vanilla-Triton capability rejection, and production Ascend-Triton dispatch; updated the docs to distinguish product images from bare and legacy environments.

Current-head quality, device-API scan, focused contracts (`8 passed, 3 skipped` plus `2 passed`), NPU-suite collection, and independent review all pass. The PR body is updated to the new head and preserves the prior 910B2 results explicitly as old-head evidence. Current-head CI has restarted.

@Luosuu The required fork workflows for current head `7a65956b` are waiting for approval. Could you please trigger them when convenient? Thanks.

**Review by Coach257 — APPROVED** [review]
